File: app/src/main/java/com/buhzzi/wuej_0/wuej_maps_activity.java
package com.buhzzi.wuej_0;

import android.view.View;

import androidx.appcompat.app.AppCompatActivity;

import com.buhzzi.wuej_0.kit.location_relative;
import com.buhzzi.wuej_0.kit.night_constants;
import com.buhzzi.wuej_0.kit.server_relative;
import com.buhzzi.wuej_0.kit.window_constants;

public class wuej_maps_activity extends AppCompatActivity {
	public wuej_maps_view maps_view;
	public int cntl_visibility = View.VISIBLE;
	public android.widget.ImageView cntl_cc_ImageView;
	public android.widget.ImageView cntl_arrow_right_ImageView;
	public android.widget.ImageView cntl_arrow_up_ImageView;
	public android.widget.ImageView cntl_arrow_left_ImageView;
	public android.widget.ImageView cntl_arrow_down_ImageView;
	public android.widget.ImageView cntl_rota_rcw_ImageView;
	public android.widget.ImageView cntl_rota_cw_ImageView;
	public android.widget.ImageView cntl_zoom_in_ImageView;
	public android.widget.ImageView cntl_zoom_out_ImageView;
	public android.widget.ImageView central_cursor_ImageView;
	public android.os.Handler locating_handler;
	@android.annotation.SuppressLint({"MissingPermission"}) public Runnable locating_runnable = () -> {
//		getSystemService(android.location.LocationManager.class).getCurrentLocation(android.location.LocationManager.GPS_PROVIDER, null, getMainExecutor(), (final android.location.Location user_loca_gps) -> {
//			if (user_loca_gps == null) {
//				getSystemService(android.location.LocationManager.class).getCurrentLocation(android.location.LocationManager.NETWORK_PROVIDER,  null, getMainExecutor(), (final android.location.Location user_loca_network) -> {
//					if (user_loca_network == null) {
//						return;
//					}
//					try {
//						update_location(wuej_maps_view.coord_in_one(user_loca_network), 0x7774656e /* "netw" */ );
//					} catch (final Exception e) { }
//				});
//				return;
//			}
//			try {
//				update_location(wuej_maps_view.coord_in_one(user_loca_gps), 0x00737067 /* "gps\0" */);
//			} catch (final Exception e) { }
//		});
		maps_view.x_in_one = wuej_maps_view.longitude_in_one(location_relative.location.getLongitude());
		maps_view.y_in_one = wuej_maps_view.latitude_in_one(location_relative.location.getLatitude());
		locating_handler.postDelayed(this.locating_runnable, 1024);
	};
	@Override public void onCreate(final android.os.Bundle saved_instance_state) {
		super.onCreate(saved_instance_state);

		server_relative.init(this);

		try {
			setContentView(R.layout.wuej_maps_activity);
			init_maps_view();
			init_cntl_views();

			(locating_handler = new android.os.Handler(android.os.Looper.getMainLooper())).post(locating_runnable);
		} catch (final Exception e) {
			e.printStackTrace();
		}
	}
	@Override public void onDestroy() {
		super.onDestroy();

		locating_handler.removeCallbacks(locating_runnable);
	}
//    @java.lang.Override public void onRequestPermissionsResult(final int request_code, final java.lang.String[] permissions, final int[] grant_results) {
//        super.onRequestPermissionsResult(request_code, permissions, grant_results);
//
//        switch (request_code) {
//            case 0x240221:
////            android.content.pm.PackageManager.
//        }
//    }
	public void init_maps_view() throws Exception {
		maps_view = findViewById(R.id.maps_view);
		if (ensure_locating_permission()) {
			@android.annotation.SuppressLint({"MissingPermission"}) final android.location.Location user_loca = getSystemService(android.location.LocationManager.class).getLastKnownLocation(android.location.LocationManager.GPS_PROVIDER);
			if (user_loca != null) {
				maps_view.set_camera_location(user_loca);
				return;
			}
		}
		maps_view.load_in_screen_maps();
	}
	public View.OnTouchListener get_cntl_on_touch(int d_pix_in_one, int d_x_in_one, int d_y_in_one, float d_rotation) {
		return (cntl_on_touch_view, cntl_on_touch_event) -> {
			cntl_on_touch_view.performClick();

			switch (cntl_on_touch_event.getAction()) {
				case android.view.MotionEvent.ACTION_DOWN:
				cntl_on_touch_view.setAlpha(0.5F);
				maps_view.d_pix_in_one = d_pix_in_one;
				maps_view.d_x_in_one = d_x_in_one;
				maps_view.d_y_in_one = d_y_in_one;
				maps_view.d_rotation = d_rotation;
				return true;

				case android.view.MotionEvent.ACTION_UP:
				cntl_on_touch_view.setAlpha(1);
				maps_view.d_pix_in_one = 0;
				maps_view.d_x_in_one = 0;
				maps_view.d_y_in_one = 0;
				maps_view.d_rotation = 0;
				return true;
			}
			return false;
		};
	}
	@android.annotation.SuppressLint({"ClickableViewAccessibility"}) public void init_cntl_views() {
		cntl_cc_ImageView = findViewById(R.id.cntl_cc_ImageView);
		cntl_cc_ImageView.setImageBitmap(draw_cntl_cc());
		cntl_cc_ImageView.setOnClickListener((final android.view.View cntl_cc_on_touch_view) -> {
			maps_view.load_in_screen_maps();
			cntl_visibility = cntl_visibility == View.VISIBLE ? View.INVISIBLE : View.VISIBLE;
			cntl_arrow_right_ImageView.setVisibility(cntl_visibility);
			cntl_arrow_up_ImageView.setVisibility(cntl_visibility);
			cntl_arrow_left_ImageView.setVisibility(cntl_visibility);
			cntl_arrow_down_ImageView.setVisibility(cntl_visibility);
			cntl_rota_rcw_ImageView.setVisibility(cntl_visibility);
			cntl_rota_cw_ImageView.setVisibility(cntl_visibility);
			cntl_zoom_in_ImageView.setVisibility(cntl_visibility);
			cntl_zoom_out_ImageView.setVisibility(cntl_visibility);
			central_cursor_ImageView.setVisibility(cntl_visibility);
			cntl_cc_ImageView.setImageBitmap(draw_cntl_cc());
		});

		cntl_arrow_right_ImageView = findViewById(R.id.cntl_arrow_right_ImageView);
		cntl_arrow_right_ImageView.setImageBitmap(draw_cntl_arrow());
		cntl_arrow_right_ImageView.setOnTouchListener(get_cntl_on_touch(0, 1, 0, 0));

		cntl_arrow_up_ImageView = findViewById(R.id.cntl_arrow_up_ImageView);
		cntl_arrow_up_ImageView.setImageBitmap(draw_cntl_arrow());
		cntl_arrow_up_ImageView.setOnTouchListener(get_cntl_on_touch(0, 0, -1, 0));

		cntl_arrow_left_ImageView = findViewById(R.id.cntl_arrow_left_ImageView);
		cntl_arrow_left_ImageView.setImageBitmap(draw_cntl_arrow());
		cntl_arrow_left_ImageView.setOnTouchListener(get_cntl_on_touch(0, -1, 0, 0));

		cntl_arrow_down_ImageView = findViewById(R.id.cntl_arrow_down_ImageView);
		cntl_arrow_down_ImageView.setImageBitmap(draw_cntl_arrow());
		cntl_arrow_down_ImageView.setOnTouchListener(get_cntl_on_touch(0, 0, 1, 0));

		cntl_rota_rcw_ImageView = findViewById(R.id.cntl_rota_rcw_ImageView);
		cntl_rota_rcw_ImageView.setImageBitmap(draw_cntl_rota());
		cntl_rota_rcw_ImageView.setOnTouchListener(get_cntl_on_touch(0, 0, 0, 1));

		cntl_rota_cw_ImageView = findViewById(R.id.cntl_rota_cw_ImageView);
		cntl_rota_cw_ImageView.setImageBitmap(draw_cntl_rota());
		cntl_rota_cw_ImageView.setOnTouchListener(get_cntl_on_touch(0, 0, 0, -1));

		cntl_zoom_in_ImageView = findViewById(R.id.cntl_zoom_in_ImageView);
		cntl_zoom_in_ImageView.setImageBitmap(draw_cntl_zoom_in());
		cntl_zoom_in_ImageView.setOnTouchListener(get_cntl_on_touch(1, 0, 0, 0));

		cntl_zoom_out_ImageView = findViewById(R.id.cntl_zoom_out_ImageView);
		cntl_zoom_out_ImageView.setImageBitmap(draw_cntl_zoom_out());
		cntl_zoom_out_ImageView.setOnTouchListener(get_cntl_on_touch(-1, 0, 0, 0));

		central_cursor_ImageView = findViewById(R.id.central_cursor_ImageView);
		central_cursor_ImageView.setImageBitmap(draw_central_cursor());
	}
	public android.graphics.Bitmap draw_cntl_cc() {
		final int size = window_constants.display_metrics.widthPixels >>> 3;
		final android.graphics.Bitmap bmp = android.graphics.Bitmap.createBitmap(
			size, size,
			android.graphics.Bitmap.Config.ARGB_8888
		);
		final boolean cntl_folded = cntl_visibility != View.VISIBLE;
		final android.graphics.Canvas cv = new android.graphics.Canvas(bmp);
		cv.drawColor(cntl_folded ? 0x40808080 : 0xc0808080);
		final android.graphics.Paint paint = new android.graphics.Paint();
		paint.setColor(night_constants.color_fore);
		if (!cntl_folded) {
			paint.setStyle(android.graphics.Paint.Style.STROKE);
			paint.setStrokeWidth(4);
		}
		cv.drawCircle(size >>> 1, size >>> 1, size * 3 >>> 3, paint);
		return bmp;
	}
	public static android.graphics.Bitmap draw_cntl_arrow() {
		final int size = window_constants.display_metrics.widthPixels >>> 3;
		final android.graphics.Bitmap bmp = android.graphics.Bitmap.createBitmap(
			size, size,
			android.graphics.Bitmap.Config.ARGB_8888
		);
		final android.graphics.Canvas cv = new android.graphics.Canvas(bmp);
		cv.drawColor(0x80808080);
		final android.graphics.Path path = new android.graphics.Path();
		path.moveTo(size >>> 2, size >>> 2);
		path.lineTo(size * 3 >>> 2, size >>> 1);
		path.lineTo(size >>> 2, size * 3 >>> 2);
		final android.graphics.Paint paint = new android.graphics.Paint();
		paint.setColor(night_constants.color_fore);
		paint.setStyle(android.graphics.Paint.Style.STROKE);
		paint.setStrokeWidth(4);
		paint.setStrokeCap(android.graphics.Paint.Cap.ROUND);
		cv.drawPath(path, paint);
		return bmp;
	}
	public static android.graphics.Bitmap draw_cntl_rota() {
		final int size = window_constants.display_metrics.widthPixels >>> 3;
		final android.graphics.Bitmap bmp = android.graphics.Bitmap.createBitmap(
			size, size,
			android.graphics.Bitmap.Config.ARGB_8888
		);
		final android.graphics.Canvas cv = new android.graphics.Canvas(bmp);
		cv.drawColor(0x40808080);
		final android.graphics.Path path = new android.graphics.Path();
		path.arcTo(size >>> 4, size * -1.9375F, size * 2.9375F, size * 0.9375F, 150, -30, true);
		path.lineTo(size >>> 2, size * 3 >>> 2);
		final android.graphics.Paint paint = new android.graphics.Paint();
		paint.setColor(night_constants.color_fore);
		paint.setStyle(android.graphics.Paint.Style.STROKE);
		paint.setStrokeWidth(4);
		paint.setStrokeCap(android.graphics.Paint.Cap.ROUND);
		cv.drawPath(path, paint);
		return bmp;
	}
	public static android.graphics.Bitmap draw_cntl_zoom_in() {
		final int size = window_constants.display_metrics.widthPixels >>> 3;
		final android.graphics.Bitmap bmp = android.graphics.Bitmap.createBitmap(
			size, size,
			android.graphics.Bitmap.Config.ARGB_8888
		);
		final android.graphics.Canvas cv = new android.graphics.Canvas(bmp);
		cv.drawColor(0x40808080);
		final android.graphics.Path path = new android.graphics.Path();
		path.moveTo(size >>> 1, size >>> 2);
		path.lineTo(size >>> 1, size * 3 >>> 2);
		path.moveTo(size >>> 2, size >>> 1);
		path.lineTo(size * 3 >>> 2, size >>> 1);
		final android.graphics.Paint paint = new android.graphics.Paint();
		paint.setColor(night_constants.color_fore);
		paint.setStyle(android.graphics.Paint.Style.STROKE);
		paint.setStrokeWidth(4);
		paint.setStrokeCap(android.graphics.Paint.Cap.ROUND);
		cv.drawPath(path, paint);
		return bmp;
	}
	public static android.graphics.Bitmap draw_cntl_zoom_out() {
		final int size = window_constants.display_metrics.widthPixels >>> 3;
		final android.graphics.Bitmap bmp = android.graphics.Bitmap.createBitmap(
			size, size,
			android.graphics.Bitmap.Config.ARGB_8888
		);
		final android.graphics.Canvas cv = new android.graphics.Canvas(bmp);
		cv.drawColor(0x40808080);
		final android.graphics.Path path = new android.graphics.Path();
		path.moveTo(size >>> 2, size >>> 1);
		path.lineTo(size * 3 >>> 2, size >>> 1);
		final android.graphics.Paint paint = new android.graphics.Paint();
		paint.setColor(night_constants.color_fore);
		paint.setStyle(android.graphics.Paint.Style.STROKE);
		paint.setStrokeWidth(4);
		paint.setStrokeCap(android.graphics.Paint.Cap.ROUND);
		cv.drawPath(path, paint);
		return bmp;
	}
	public static android.graphics.Bitmap draw_central_cursor() {
		final int size = window_constants.display_metrics.widthPixels >>> 4;
		final android.graphics.Bitmap bmp = android.graphics.Bitmap.createBitmap(
			size, size,
			android.graphics.Bitmap.Config.ARGB_8888
		);
		final android.graphics.Canvas cv = new android.graphics.Canvas(bmp);
		cv.drawColor(0x00000000);
		final android.graphics.Path path = new android.graphics.Path();
		path.moveTo(size >>> 1, 0);
		path.lineTo(size >>> 1, size * 3 >>> 3);
		path.moveTo(size >>> 1, size * 5 >>> 3);
		path.lineTo(size >>> 1, size);
		path.moveTo(0, size >>> 1);
		path.lineTo(size * 3 >>> 3, size >>> 1);
		path.moveTo(size * 5 >>> 3, size >>> 1);
		path.lineTo(size, size >>> 1);
		final android.graphics.Paint paint = new android.graphics.Paint();
		paint.setColor(0xffff0000);
		paint.setStyle(android.graphics.Paint.Style.STROKE);
		paint.setStrokeWidth(4);
		paint.setStrokeCap(android.graphics.Paint.Cap.ROUND);
		cv.drawPath(path, paint);
		return bmp;
	}
	public boolean ensure_locating_permission() {
		if (androidx.core.app.ActivityCompat.checkSelfPermission(this, android.Manifest.permission.ACCESS_FINE_LOCATION) == android.content.pm.PackageManager.PERMISSION_GRANTED) {
			return true;
		}
		androidx.core.app.ActivityCompat.requestPermissions(this, new String[]{android.Manifest.permission.ACCESS_FINE_LOCATION}, 0x24022116);
		return false;
	}
	public void update_location(final long coord_one, final int type) { }
}

File: app/src/main/java/com/buhzzi/wuej_0/WuejMapsActivity.kt
package com.buhzzi.wuej_0

import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.location.Location
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.core.app.ActivityCompat
import com.buhzzi.wuej_0.kit.ByteRelative
import com.buhzzi.wuej_0.kit.FontConstants
import com.buhzzi.wuej_0.kit.LocationConverter
import com.buhzzi.wuej_0.kit.LocationRelative
import com.buhzzi.wuej_0.kit.OrientationRelative
import com.buhzzi.wuej_0.kit.RequestHelper
import com.buhzzi.wuej_0.kit.RsaRelative
import com.buhzzi.wuej_0.kit.ServerRelative
import com.buhzzi.wuej_0.kit.SocketRequestHelper
import com.buhzzi.wuej_0.kit.StackedActivity
import com.buhzzi.wuej_0.kit.StreamHelper
import com.buhzzi.wuej_0.kit.WuejMapsDrawers
import com.buhzzi.wuej_0.kit.WuejMapsView
import org.json.JSONObject
import java.util.Base64
import kotlin.io.path.div
import kotlin.io.path.exists
import kotlin.io.path.readBytes
import kotlin.math.max

class WuejMapsActivity : StackedActivity() {
	private var firstTime = true
	private lateinit var mapsView: WuejMapsView
	private var cntlVisibility = View.VISIBLE
	private lateinit var cntlCcImg: ImageView
	private lateinit var cntlArrowRightImg: ImageView
	private lateinit var cntlArrowUpImg: ImageView
	private lateinit var cntlArrowLeftImg: ImageView
	private lateinit var cntlArrowDownImg: ImageView
	private lateinit var cntlRotaRcwImg: ImageView
	private lateinit var cntlRotaCwImg: ImageView
	private lateinit var cntlZoomInImg: ImageView
	private lateinit var cntlZoomOutImg: ImageView
	private lateinit var centralCursorImg: ImageView
	private lateinit var infoLay: LinearLayout
	private lateinit var longiTxt: TextView
	private lateinit var latiTxt: TextView
	private val locatingHandler = Handler(Looper.getMainLooper())
	private lateinit var locatingRunnable: Runnable

	private lateinit var xuh: String
	private lateinit var lorif: ByteArray
	private lateinit var locup: ByteArray
	private lateinit var chaschig: Array<Array<String>>

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		if (ensureLocatingPermission()) {
			initWithLocatingPermission()
		}
	}

	override fun onDestroy() {
		super.onDestroy()

		if (::locatingRunnable.isInitialized) {
			locatingHandler.removeCallbacks(locatingRunnable)
		}

		LocationRelative.stopUpdatingLocation()
		OrientationRelative.stopUpdatingOrientation()
	}

	override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String?>, grantResults: IntArray) {
		super.onRequestPermissionsResult(requestCode, permissions, grantResults)

		if (requestCode == REQUEST_LOCATING_PERMISSION_CODE) {
			if (
				grantResults.isNotEmpty() &&
				grantResults[0] == PackageManager.PERMISSION_GRANTED
			) {
				initWithLocatingPermission()
			}
		}
	}

	private fun initWithLocatingPermission() {
		setContentView(R.layout.wuej_maps_activity)
		initMapsView()
		initXuh()
		initCntlViews()
		initInfo()

		// You should turn off background battery saver to continuously call this lambda.
		locatingRunnable = Runnable {
			locatingHandler.postDelayed(locatingRunnable, 512)
			LocationRelative.location?.let {
				if (firstTime) {
					firstTime = false
					locateMyself()
				}
				mapsView.loadInScreenMaps()
				longiTxt.text = LocationConverter.longitudeToFrac(it.longitude)
				latiTxt.text = LocationConverter.latitudeToFrac(it.latitude)

				if (::lorif.isInitialized && ::locup.isInitialized) {
					setAndGetLocations(it)
				}
			} ?: run {
				longiTxt.text = getString(R.string.maps_unknown_longitude)
				latiTxt.text = getString(R.string.maps_unknown_latitude)
			}
			centralCursorImg.rotation = OrientationRelative.azimuth
		}
		locatingHandler.post(locatingRunnable)

		LocationRelative.startUpdatingLocation()
		OrientationRelative.startUpdatingOrientation()
	}

	private fun initMapsView() {
		mapsView = findViewById(R.id.mapsView)
	}

	private fun getCntlOnTouch(dPixInOne: Int, dXInOne: Int, dYInOne: Int, dRotation: Float): (View, MotionEvent) -> Boolean {
		return { cntlOnTouchView, cntlOnTouchEvent ->
			cntlOnTouchView.performClick()
			when (cntlOnTouchEvent.action) {
				MotionEvent.ACTION_DOWN -> {
					cntlOnTouchView.alpha = 0.5F
					mapsView.dPixInOne = dPixInOne
					mapsView.dXInOne = dXInOne
					mapsView.dYInOne = dYInOne
					mapsView.dRotation = dRotation
					true
				}

				MotionEvent.ACTION_UP -> {
					cntlOnTouchView.alpha = 1F
					mapsView.dPixInOne = 0
					mapsView.dXInOne = 0
					mapsView.dYInOne = 0
					mapsView.dRotation = 0F
					true
				}

				else -> false
			}
		}
	}

	@SuppressLint("ClickableViewAccessibility")
	private fun initCntlViews() {
		cntlCcImg = findViewById(R.id.cntlCcImageView)
		cntlCcImg.setImageBitmap(WuejMapsDrawers.drawCntlCc(cntlVisibility))
		cntlCcImg.setOnClickListener {
			mapsView.loadInScreenMaps()
			cntlVisibility = if (cntlVisibility == View.VISIBLE) View.INVISIBLE else View.VISIBLE
			cntlCcImg.setImageBitmap(WuejMapsDrawers.drawCntlCc(cntlVisibility))
			cntlArrowRightImg.visibility = cntlVisibility
			cntlArrowUpImg.visibility = cntlVisibility
			cntlArrowLeftImg.visibility = cntlVisibility
			cntlArrowDownImg.visibility = cntlVisibility
			cntlRotaRcwImg.visibility = cntlVisibility
			cntlRotaCwImg.visibility = cntlVisibility
			cntlZoomInImg.visibility = cntlVisibility
			cntlZoomOutImg.visibility = cntlVisibility
			centralCursorImg.visibility = cntlVisibility
			infoLay.visibility = cntlVisibility
		}
		cntlCcImg.setOnLongClickListener {
			locateMyself()
			true
		}
		cntlArrowRightImg = findViewById(R.id.cntlArrowRightImageView)
		cntlArrowRightImg.setImageBitmap(WuejMapsDrawers.drawCntlArrow())
		cntlArrowRightImg.setOnTouchListener(getCntlOnTouch(0, 1, 0, 0f))

		cntlArrowUpImg = findViewById(R.id.cntlArrowUpImageView)
		cntlArrowUpImg.setImageBitmap(WuejMapsDrawers.drawCntlArrow())
		cntlArrowUpImg.setOnTouchListener(getCntlOnTouch(0, 0, -1, 0f))

		cntlArrowLeftImg = findViewById(R.id.cntlArrowLeftImageView)
		cntlArrowLeftImg.setImageBitmap(WuejMapsDrawers.drawCntlArrow())
		cntlArrowLeftImg.setOnTouchListener(getCntlOnTouch(0, -1, 0, 0f))

		cntlArrowDownImg = findViewById(R.id.cntlArrowDownImageView)
		cntlArrowDownImg.setImageBitmap(WuejMapsDrawers.drawCntlArrow())
		cntlArrowDownImg.setOnTouchListener(getCntlOnTouch(0, 0, 1, 0f))

		cntlRotaRcwImg = findViewById(R.id.cntlRotaRcwImageView)
		cntlRotaRcwImg.setImageBitmap(WuejMapsDrawers.drawCntlRota())
		cntlRotaRcwImg.setOnTouchListener(getCntlOnTouch(0, 0, 0, 1f))

		cntlRotaCwImg = findViewById(R.id.cntlRotaCwImageView)
		cntlRotaCwImg.setImageBitmap(WuejMapsDrawers.drawCntlRota())
		cntlRotaCwImg.setOnTouchListener(getCntlOnTouch(0, 0, 0, -1f))

		cntlZoomInImg = findViewById(R.id.cntlZoomInImageView)
		cntlZoomInImg.setImageBitmap(WuejMapsDrawers.drawCntlZoomIn())
		cntlZoomInImg.setOnTouchListener(getCntlOnTouch(1, 0, 0, 0f))

		cntlZoomOutImg = findViewById(R.id.cntlZoomOutImageView)
		cntlZoomOutImg.setImageBitmap(WuejMapsDrawers.drawCntlZoomOut())
		cntlZoomOutImg.setOnTouchListener(getCntlOnTouch(-1, 0, 0, 0f))

		centralCursorImg = findViewById(R.id.centralCursorImageView)
		centralCursorImg.setImageBitmap(WuejMapsDrawers.drawCentralCursor())
	}

	private fun initInfo() {
		infoLay = findViewById(R.id.infoLinearLayout)
		val infoTxtList = listOf(
			getString(R.string.maps_unknown_longitude),
			getString(R.string.maps_unknown_latitude),
			SettingActivity.providerNames[SettingActivity.provider],
			SettingActivity.minTimeNames[SettingActivity.minTime],
			SettingActivity.minDistanceNames[SettingActivity.minDistance],
		).map { name ->
			TextView(this).apply {
				layoutParams = LinearLayout.LayoutParams(
					ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT,
				)
				setBackgroundColor(0x80000000.toInt())
				text = name
				typeface = FontConstants.xeuTf
				setTextColor(0xffffffff.toInt())
				textSize = 16F
				infoLay.addView(this)
			}
		}
		longiTxt = infoTxtList[0]
		latiTxt = infoTxtList[1]
	}

	private fun initXuh() {
		(filesDir.toPath() / "wuej_xuh")
			.takeIf { it.exists() }
			?.readBytes()
			?.run {
				val nmLn = (this[0].toUInt() or (this[1].toUInt() shl 8)).toInt()
				xuh = decodeToString(2, nmLn + 2)
				val otLn = this[nmLn + 2].toUByte().toInt() + 1024
				lorif = copyOfRange(nmLn + 3, otLn + nmLn + 3)
				locup = copyOfRange(otLn + nmLn + 3, size)
			}
		chaschig = run {
			(filesDir.toPath() / "chaschig")
				.takeIf { it.exists() }
				?.readBytes()
				?: "{}".encodeToByteArray()
		}.run { // hashes
			JSONObject(decodeToString())
		}.run { keys().asSequence().map { k -> arrayOf(k, getString(k)) }.toList().toTypedArray() }
	}

	private fun setAndGetLocations(location: Location) = (arrayOf(
		WuejMapsView.longitudeInOne(location.longitude).toUInt(),
		WuejMapsView.latitudeInOne(location.latitude).toUInt(),
	) + OrientationRelative.orientation.map { f -> f.toRawBits().toUInt() })
		.joinToString(",")
		.toByteArray()
		.run { RsaRelative.encrypt(locup, this) }
		.run {
			Base64.getEncoder().encodeToString(this)
				.apply { print(this) }
		}
		.run {
			mapOf(
				"task" to "location_240830_set_and_get",
				"name" to chaschig.first { p -> p[1] == xuh }[0],
				"loca" to this,
			)
		}
		.run { JSONObject(this) }
		.toString()
		.toByteArray()
		.run {
			RsaRelative.encrypt(
				resources.openRawResource(R.raw._240828)
					.run { StreamHelper.readInputStream(this) },
				this,
			)
		}
		.run {
			ServerRelative.magic +
				ByteRelative.intToBytes(this.size + 16) +
				ServerRelative.jsonMagic +
				this
		}
		.run {
			SocketRequestHelper.request(
				String(ServerRelative.getTshr16()),
				SettingActivity.serverAddr + "/",
				emptyList(),
				this,
				(RequestHelper.json_callback { code, _, body ->
					runOnUiThread {
						code == 200 || return@runOnUiThread
						mapsView.contactNames = body.keys().asSequence().map { k ->
							chaschig.first { m -> m[0] == k }[1]
						}.toList().toTypedArray()
						mapsView.contactLocaArr = body.keys().asSequence().map { chasch ->
							val (xInOne, yInOne) = body.getString(chasch)
								.run { Base64.getDecoder().decode(this) }
								.run { RsaRelative.decrypt(lorif, this) }
								.decodeToString()
								.split(',')
							((xInOne.toULong() shl 32) or yInOne.toULong()).toLong()
						}.toList().toTypedArray().toLongArray()
					}
				}).toCallback(),
			)
		}

	private fun ensureLocatingPermission(): Boolean {
		if (
			ActivityCompat.checkSelfPermission(
				this, Manifest.permission.ACCESS_FINE_LOCATION,
			) == PackageManager.PERMISSION_GRANTED
		) {
			return true
		}
		ActivityCompat.requestPermissions(
			this,
			arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
			REQUEST_LOCATING_PERMISSION_CODE,
		)
		return false
	}

	private fun locateMyself() = LocationRelative.location?.let {
		mapsView.setCameraLocation(it)
		mapsView.pixInOne = max(mapsView.pixInOne, 0x20000.toDouble())
	} ?: run {
		Toast.makeText(this, R.string.maps_location_unavailable, Toast.LENGTH_SHORT).show()
	}

	companion object {
		private const val REQUEST_LOCATING_PERMISSION_CODE = 0x24022116
	}
}
